### ISSUE TITLE
DesktopGL KeyDown and KeyUp event potential performance improvements

### ISSUE THREAD

**Issue by tdeeb** [MonoGame/MonoGame#7035]
`SDLGamePlatform` maintains a list of unique `Keys` that are pressed. This list is [shared with `Keyboard`](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Platform/Input/Keyboard.SDL.cs#L11) and passed into newly created `KeyboardStates`, where it is iterated upon to set the key flags.

Whenever a new key down event is processed by SDL, a [`Contains()` check](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Platform/SDL/SDLGamePlatform.cs#L138) is in place to prevent adding duplicates to this `Keys` list. Similarly, [`Remove()` is called](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Platform/SDL/SDLGamePlatform.cs#L149) in the key up event; `Remove()` internally iterates through the list to find if the value actually exists before removing it.

Since all keys in this list are unique and the order of the list does not matter, I propose this collection to be changed from a `List<Keys>` to a `HashSet<Keys>` to improve performance of the `Contains()` and `Remove()` calls from O(n) to O(1). Iterating through a `HashSet` instead of a `List` will not change the result of `KeyboardStates` created since, again, the order of the flags set does not matter.

#### What version of MonoGame does the bug occur on:
- Develop

#### What operating system are you using:
- Windows, Linux

#### What MonoGame platform are you using:

- DesktopGL

**Comment by harry-cpp** [comment]
Correct, however, you can optimise it even further by doing just `|=` and `&=` operations with a local KeyboardState.

On a related note, I noticed this before but kinda forgot about it during the SDL switch. Reason being is Keyboard.Default.cs takes a List as one of its params and to fix this this you'd need to go over multiple platforms fixing the same problem.

**Comment by tamkcr** [comment]
Last I remember HashSet creates a lot of garbage, would avoid it like plague in input code.

**Comment by willmotil** [comment]
The current implementations make only some start up garbage but no deallocations. That would cause eventual collections there after. 

Make sure to check the allocations.

**Comment by tdeeb** [comment]
To maximize compatibility, what if the list remains so the keyboard can use it and DesktopGL has a Dictionary or something similar to check for duplicates?

**Comment by Jjagg** [comment]
We should fix all platforms at once, there's no need to keep compatibility with our current implementation. And it should use KeyboardState like Harry said, since it's optimized for representing what keys are down. It'll be faster and lighter than a HashSet.

**Comment by harry-cpp** [comment]
> But really this is very very very minor. I mean how many keys are in the list at most, like 5?

Most keyboards support up to 4 simultaneous keys iirc.

**Comment by tamkcr** [comment]
Biologically at max we have 9 other fingers. A rough throwaway test (4.7) I did shows in such worst case `HashSet` is ~33% faster than `List`, and in-differentiable against an array. Note that the speed gain is only there with `int`s, with `Keys` in `HashSet` is about 3 times slower, I suspect its the boxing cost from `GetHashCode`.

In .net core where I remember reading they mitigated the enum boxing, well I couldn't get it to work properly as `HashSet` appeared to be 10times slower than a `List` so I probably messed up something.

In any case, I don't personally think this is worth changing, the difference is minuscule.

**Comment by DameonL** [comment]
With small collections, a HashSet performs worse than a List for Contains, even though Contains is O(1) for a HashSet and O(n) for a List. This is because time complexity only measures how the computation cost scales as the collection grows. If you only have 3-10 entries, then the cost of calculating a hash is higher than the cost of just going through 3 array entries and doing a comparison, most of the time.
